src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -254,4 +267,106 @@ void justClose() {
             SCOPED_MEMORY_ACCESS.closeScope(this);
         }
     }
+
+    /**
+     * A functional interface modelling the cleanup action associated with a scope.
+     */
+    interface CleanupAction extends Runnable {
+        void cleanup();
+        CleanupAction dup();
+
+        @Override
+        default void run() {
+            cleanup();
+        }
+
+        /** Dummy cleanup action */
+        CleanupAction DUMMY = new CleanupAction() {
+            @Override
+            public void cleanup() {
+                // do nothing
+            }
+
+            @Override
+            public CleanupAction dup() {
+                return this;
+            }
+        };
+
+        /**
+         * A stateful cleanup action; this action can only be called at most once. The implementation
+         * guarantees this invariant even when multiple threads race to call the {@link #cleanup()} method.
+         */
+        abstract class AtMostOnceOnly implements CleanupAction {
+
+            static final VarHandle CALLED;
+
+            static {
+                try {
+                    CALLED = MethodHandles.lookup().findVarHandle(AtMostOnceOnly.class, "called", boolean.class);
+                } catch (Throwable ex) {
+                    throw new ExceptionInInitializerError(ex);
+                }
+            }
+
+            private boolean called = false;
+
+            abstract void doCleanup();
+
+            public final void cleanup() {
+                if (disable()) {
+                    doCleanup();
+                }
+            };
+
+            @Override
+            public CleanupAction dup() {
+                disable();
+                class DupAction extends AtMostOnceOnly {
```

Comment:
Nice approach. Shame we cannot declare static classes within a method, since the inner class does not need to implicitly capture "this" of the outer class. I suppose we could make it static for some marginal efficiency gain reducing dependency chains.